### PR TITLE
issue: 4712119 Restrict SO_XLIO_EXT_VLAN_TAG to 12 bits

### PR DESCRIPTION
**Pull request by pasis**
## Description
SO_XLIO_EXT_VLAN_TAG option inserts VLAN tag to TX packets on the configured socket. Current requirement for the feature is to support the VLAN ID part of the VLAN tag.

Return EINVAL if the option value is >= than 4096.

##### What
Return EINVAL if the option value is >= than 4096.

##### Why ?
Guard against unexpected input.

## Change type
What kind of change does this PR introduce?
- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] Tests
- [ ] Other

## Check list
- [ ] Code follows the style de facto guidelines of this project
- [ ] Comments have been inserted in hard to understand places
- [ ] Documentation has been updated (if necessary)
- [ ] Test has been added (if possible)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed validation for the extended VLAN tag socket option to correctly accept only valid 12-bit VLAN ID values (0-4095), rejecting out-of-range inputs.

* **Tests**
  * Added comprehensive test coverage for VLAN tag socket option validation, including edge cases and error handling.

### CODE REVIEW

**Comment by tomerdbz** [comment]
make sure to add to release notes - as this breaks backwards compat
DEI bit lost
priority should be set via SO_PRIORITY

File: tests/gtest/tcp/tcp_sockopt.cc [hunk 1]
```diff
@@ -216,6 +216,33 @@ TEST_F(tcp_sockopt, ti_2_tcp_congestion)
...
+    int val = 4095;
+    int rc = setsockopt(fd, SOL_SOCKET, SO_XLIO_EXT_VLAN_TAG, &val, sizeof(val));
+    EXPECT_EQ(0, rc);
+
+    val = 4096;
+    rc = setsockopt(fd, SOL_SOCKET, SO_XLIO_EXT_VLAN_TAG, &val, sizeof(val));
+    EXPECT_EQ(-1, rc);
+    EXPECT_EQ(EINVAL, errno);
+
+    close(fd);
```
[the remark this answers is not in the document]
Comment (tomerdbz) [reply tests/gtest/tcp/tcp_sockopt.cc:253]: test:
tempval >= 0 guard - negative (as greptile noted). val = -1 should give EINVAL.
__optlen == sizeof(int) guard - bad optlen not tested (e.g. sizeof(short)) → EINVAL.
val = 0 accept path (the "fall back to dev_vlan" case) - untested.

File: src/core/sock/sockinfo_tcp.cpp [hunk 2]
```diff
@@ -4869,7 +4869,8 @@ int sockinfo_tcp::tcp_setsockopt(int __level, int __optname, __const void *__opt
         case SO_XLIO_EXT_VLAN_TAG:
             if (__optlen == sizeof(int)) {
                 int tempval = *reinterpret_cast<const int *>(__optval);
-                if (tempval >= 0 && tempval <= UINT16_MAX) {
+                /* Option supports 12-bit VLAN ID part of the VLAN tag. */
+                if (tempval >= 0 && tempval < 4096) {
```
Comment (tomerdbz) [inline src/core/sock/sockinfo_tcp.cpp:4873]: From the RFC:

Image

any deviations need to be documented/explained. See our conv regarding 0. Also note 0xFFF

**Review by tomerdbz — CHANGES_REQUESTED** [review]

File: src/core/sock/sockinfo_tcp.cpp (hunk 2 above)
Comment (pasis) [reply src/core/sock/sockinfo_tcp.cpp:4873]: value 0 is documented in a comment and covered with  test. Its functionality remains the same: disable previously configured feature.

Values 1 and 0xFFF are not restricted and this is documented.

File: tests/gtest/tcp/tcp_sockopt.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (pasis) [reply tests/gtest/tcp/tcp_sockopt.cc:253]: added extra tests

File: src/core/sock/sockinfo_tcp.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pasis) [reply src/core/sock/sockinfo_tcp.cpp:4873]: I prefer to keep raw number in this specific case

**Comment by pasis** [comment]
> make sure to add to release notes - as this breaks backwards compat DEI bit lost priority should be set via SO_PRIORITY

This is very specific feature and its purpose to provide information in TX path rather than to configure fully compliant 802.1Q support. In current state this feature keeps compatibility with 802.1Q tag, but this can be changed in the future if needed. Note, the option doesn't allow to build 2-way communication as is, because RX path is not handled.

**Review by tomerdbz — DISMISSED** [review]

**Review by tomerdbz — APPROVED** [review]
